### ISSUE TITLE
Search bar for apps and key buttons for volume

### ISSUE THREAD

**Issue by D3vil0p3r** [end-4/dots-hyprland#12]
Hey mate.

I'm testing your Summer Gruv environment and it seems so cool!

The search bar accessible on the top-center of the top bar, when I try to write an application I installed, for example kitty or brave, it does not find anything. Of course if I type the complete command, it runs the app, but I expected it could show the list of the applications I was searching. Do I need to configure it in order to find the installed applications?
![image](https://github.com/end-4/dots-hyprland/assets/83867734/ee2e779b-3080-470b-bb0d-24cc88d07175)

At the end, on my keyboard, usually, for increasing and decreasing volume I use `Fn + LeftArrow` and `Fn + RightArrow` keys. When I press them, I can see the volume icon appearing on the screen but it does not increase or decrease:
![image](https://github.com/end-4/dots-hyprland/assets/83867734/ffdbfdfb-7d45-4851-b70d-74df7be42508)

**Comment by end-4** [comment]
### Search
- Attach `~/.config/eww/scripts/cache/entries.txt` and `~/.config/eww/scripts/cache/entrynames.txt` pls.
- If the file already exists with content, good. Try running `cd ~/.config/eww && scripts/appsearch 'kit' | gojq` and see if it prints a json array with matching apps.
### Volume OSD
- Try pressing the key again and see if it works. Eww + Bash is slow so this happens sometimes
- If that still doesn't work, try manually running `cd ~/.config/eww && scripts/volume` and see if it outputs a json normally

**Comment by D3vil0p3r** [comment]
About **Search**, there is not a `cache` directory in `~/.config/eww/scripts/`.
```
cd ~/.config/eww && scripts/appsearch 'kit' | gojq

cat: scripts/cache/entrynames.txt: No such file or directory
cat: scripts/cache/entries.txt: No such file or directory
[]
```

About **Volume OSD**, by pressing it again and keeping it, no work, only the Volume icon appears. Btw, if I use `Fn + UpArrow` and `Fn + DownArrow` for increasing and decreasing the screen brightness, it works.
The last command you asked has this output:
```
cd ~/.config/eww && scripts/volume                            🬈🬖🬥🬅 time:48ms 🬖🬥🬔🬗
Could not connect to PipeWire
Could not connect to PipeWire
Could not connect to PipeWire
Could not connect to PipeWire
Could not connect to PipeWire
{"icon":"","audio":"1","percent":"","microphone":""}
```
Pipewire is installed with the version `pipewire-1:0.3.71-2` that should be the latest in Arch.

Another "typo" I detected, maybe, is that, on the **Search** there is written **LMB: Focus | MMB: Close | RMB: Select/Move**, but I noted that maybe MMB and RMB text must be exchanged because I note that the Search closes when I press RMB and not MMB.

**Comment by end-4** [comment]
### Search
- You should create that directory manually for now. I'll make the script mkdir that after this
### Volume
- Looks like it's your own problem. Unfortunately I have never installed raw Arch so I don't know how to handle this kinda stuff.
- Does that script print the json every time you change volume?
### The overview typo
- Not typo. The described actions are for windows. For example, this is how you move a window to another workspace using RMB:

https://github.com/end-4/dots-hyprland/assets/97237370/fc012d41-0fe1-43e3-89db-59d1f1faac42

**Comment by D3vil0p3r** [comment]
For **Search**, I created the `cache` folder and inside it I can see `kitty` and `brave-browser`. What I note is that, Brave (on the left) has no icon despite it is stored in `/usr/share/icons/hicolor/MxN/apps/brave-desktop.png`:
![image](https://github.com/end-4/dots-hyprland/assets/83867734/ea86f172-8df3-482c-ac78-bf445a7fb8ed)

And for what I understand, that search bar does not show all the installed apps starting with the string I'm typing, but I should write directly the name of the app, right?

About the volume, each time I run that script and I click on `Fn + Left or RightArrow`, I get this:
```
Could not connect to PipeWire
Could not connect to PipeWire
Could not connect to PipeWire
Could not connect to PipeWire
Could not connect to PipeWire
{"icon":"","audio":"1","percent":"","microphone":""}

Tried to close window named 'winosd_bright', but no such window was openTried to close window named 'winosd_vol', but no such window was open

Tried to close window named 'winosd_bright', but no such window was open
Tried to close window named 'winosd_vol', but no such window was open
Tried to close window named 'osd', but no such window was open
Tried to close window named 'winosd_vol', but no such window was open
Tried to close window named 'winosd_bright', but no such window was open

Tried to close window named 'winosd_bright', but no such window was open
Tried to close window named 'winosd_vol', but no such window was open
Tried to close window named 'winosd_bright', but no such window was open
Tried to close window named 'winosd_vol', but no such window was open
```
That `Tried to close window` message appears.

**Comment by end-4** [comment]
### The icon
- In the overview, hover on that Brave window and see what class name appears.
- Run `geticons CLASS_NAME` and see if it returns any result. Replace `CLASS_NAME` with that class name you found above.

### Search
- It's quite basic a basic search. It only searches apps based only on their names, not type/description/keyword.
- It uses fuzzy matching. If you dig in the script you'll see that it uses `fzf`. 
- It shows only 10 first results. Eww is slow and I don't want it to show too many apps (especially when you start typing, there will be a hella lot of results)

### Volume
- Does the OSD show up?
- Try to silent the output to see if it fixes things (imma do that now)

### Your pipewire problem
- Alright, good that it worked!

**Comment by D3vil0p3r** [comment]
> ### The icon
> * In the overview, hover on that Brave window and see what class name appears.
> * Run `geticons CLASS_NAME` and see if it returns any result. Replace `CLASS_NAME` with that class name you found above.

`CLASS_NAME` that appears is `Brave-browser`. When I run `geticons Brave-browser` I don't get any output.

> 
> ### Search
> * It's quite basic a basic search. It only searches apps based only on their names, not type/description/keyword.
> * It uses fuzzy matching. If you dig in the script you'll see that it uses `fzf`.
> * It shows only 10 first results. Eww is slow and I don't want it to show too many apps (especially when you start typing, there will be a hella lot of results)

The problem is that I don't get any result when I start to type an application that I have, for example if I start to type `foot` for the Foot terminal, I don't get any results:
![image](https://github.com/end-4/dots-hyprland/assets/83867734/6b99c879-b587-47c5-a8ab-d65f7b44155c)


> 
> ### Volume
> * Does the OSD show up?
> * Try to silent the output of the eww close command to see if it fixes things (imma do that now)

Solved by pipewire commands reported above.

PS: meanwhile, in your next commit, could you please `mkdir .config/eww/scripts/cache`?

**Comment by end-4** [comment]
### The icon
- The icon pack doesn't seem to support brave. What you can do is edit `index.theme` file of the icon pack so that the `Inherits=` section lists another icon pack that supports brave.
### Search 
- Please tell me if `~/.config/eww/scripts/cache/entries.txt` and `~/.config/eww/scripts/cache/entrynames.txt` exist. Check if they're a JSON and a list of applications, respectively, or are files of error messages.
- If the two files are fine, try running `cd ~/.config/eww/ && scripts/appsearch 'foo'`. See if it outputs a valid JSON.
### mkdir and username
- well sure

**Comment by D3vil0p3r** [comment]
For **Icon**, where is the path of `index.theme`?

For **Search**, I have those two txt files but they are empty.

One big doubt: I see you committed the creation of `cache` folder in main branch. I'm using `summergruv` branch and I note that the commits you do in the main are not replicated in summergruv. Could be that summergruv branch could have old (and maybe) bugged files?
If so, is it possible to keep in the main branch those config files that are equal for all sub-branches and having files of the sub-branches all updated?

**Comment by end-4** [comment]
### Icons
- `index.theme` is located in the root of an icon pack. That is, in this case, `.local/share/icons/Win11-dark/index.theme`
### Search
- The 2 text files are empty, meaning that the app listing script failed. Run them manually (`cd ~/.config/eww && scripts/allappnames` and `cd ~/.config/eww && scripts/allapps`) to see if there are any errors.
### Branches
- As noted in the readme, non-main branches are archives. Therefore they may have fewer features than the main branch, as it's the only one that I actively work on. 
- I should make it clear that (for now) it's better to clone the main branch, as it's newer and has fewer bugs. `summer-gruv` is there for people who also want gruvbox colors, but if you're coming here mostly for my eww config, you should use the main branch.
### summer-gruv
- Theme: What? I'm quite sure I uploaded the gruvbox one... I copied the gruvbox theme to the repo again and gh desktop didnt show changes... Have you tried logging out and logging in? (blame gtk4). Only gtk4 theme is available, the repo does not contain a gtk3 gruvbox theme, ~~I forgor~~ okay bundled https://github.com/end-4/dots-hyprland/commit/9a270e5de20653326619fd8fc004ee5e2de43abd
- The reload button: it loads the default theme. Maybe by doing that it reloaded stuff but didnt close windows correctly so the themer window remains open, hogging mouse focus. Reload eww and it should be fine.
- Icon: it is included. It's Bibata Modern Classic and it's in the [icons folder](https://github.com/end-4/dots-hyprland/tree/summer-gruv/.local/share/icons)

**Comment by D3vil0p3r** [comment]
### Icons
* Does it work also for icons in `/usr/share/icons` folder?

### Search
* The error comes when I run `allappnames` because it goes to search for `/var/lib/flatpak/exports/share/applications/` directory, but this directory does not exist by default on Arch Linux distros. In the C++ source file, could you please add a control that, before running the following code lines:
  ```
  get_desktop_entries("/usr/share/applications/");
  get_desktop_entries("/home/" + username + "/.local/share/applications");
  get_desktop_entries("/var/lib/flatpak/exports/share/applications");
  ```
  it checks if each one of these directories exist? In this manner the error should not be produced in case one of these dirs does not exist, so this script could work in every distro.
* about `allapps` script, could you edit the following lines:
  ```
  string path = readIfExists("scripts/cache/" + iconname);
  ```
  and
  ```
  writeToFile("scripts/cache/" + iconname, path);
  ```
  in order to contain the entire absolute path as `/home/ + username + /.config/eww/scripts/cache`? In this manner,  the code is not more dependent from the folder that `appnames` is running from.

### Branches
* Since the `main` branch is the only one with the right files, and since other branches like `summer-gruv` have old files but also files related to their specific theme (i.e., `.themes/Gruvbox`, fish or terminal config files for the specific theme, and so on), could you please remove the useless old config files from those branches and keep only those themed files that I can use to replace the `main` branch themed files in order to apply `summer-gruv` or other themes easily?

### Freezing environment buttons
* I note that the problem of freezing due to the click on `reload`, `music` (and I discovered also the top-left gear icon button on the top bar) is occurring also on the `main` branch. I cannot reload eww from the environment because I cannot click or focus to the terminal for writing `eww reload`. The only action that I can do is to call a TTY by `CTRL+ALT+Fx` and, from there, running `eww reload` or `reboot`, but it is uncomfortable for a user. Is there a chance to fix this issue?

**Comment by end-4** [comment]
### Icons
- The location I mentioned is for the icon pack in the repo. You should copy the icon pack to `/usr/share/icons` for Search to work.
### Search
- I see. Will fix that.
- App search dir: Okay.. can do that. I just don't see why. You aren't supposed to run that outside of the eww folder anyway. What I do is have another app search available in case my eww one fails.
### Branches
- Well yeah, it really is kind of a mess right now. I will clean it... some time in the future. Maybe I can use tags to keep that gruvbox setup :thinking: 
### Freezing buttons hogging mouse clicks
- Welp, welcome to eww! lol. What you can do is stop abusing power savers and not use the xanmod kernel (im not entirely sure but those seem to cause such problems for me)
- A quicker way to restart eww is to launch 2 terminals, focus one and reload/kill eww entirely using your keyboard and keybinds. This is how I do it:
  - In `hyprland.conf`:
  ```ini
  bind = ALT, Tab, cyclenext
  bind = SUPER, T, exec, foot
  ```
  - Press `Super`+`T` twice. Press `Alt`+`Tab`. Type the command and hit enter.

**Comment by D3vil0p3r** [comment]
@end-4  thank you man. Sorry if I asked all these small fixes, it is because I would like to implement your Hyprland theme in the next release of [Athena OS](https://github.com/Athena-OS/athena-iso).

About the freezinfg buttons, the linux kernel is the zen one. I will try to give a look on the code. Could you please tell me where are the pieces of code managing the event of clicking both the music and reload buttons and the button on top-left of the top bar? I will try to give a look.

About the other points, keep in touch when you cleaned the branches (I'm creating an Arch package for this theme).

**Comment by end-4** [comment]
Fixed the script
https://github.com/end-4/dots-hyprland/commit/5beeebc98d9edf18d7bbc8a4cd2a5dc2afab433e

I guess I imma do the branch cleanup when I'm done with my new rice

**Comment by D3vil0p3r** [comment]
Thank you mate. Meanwhile I could ask you if you could tell me where are the files with the pieces of code managing the event of clicking both the music and reload buttons so I can give a look?

**Comment by end-4** [comment]
### The freezing problem
- The theme loader window is internally still open, which makes everything unclickable. Clicking outside should close it. No?
- Commenting those lines make it not freeze might be because it makes the command run faster and therefore not forcibly stopped by eww. I should add a `&` at the end of the onclick command of the button to fix that.
### The default theme
- Oh. I'll just copy some theme there.
### Icons
- I actually do use win11 icons (they're clean, I legitimately like it). I don't want to bundle many icons packs as it will just increase the repo size. This is a dotfiles repo, not an icon collection.

**Comment by end-4** [comment]
I fixed the themer freezing issue by making those buttons close that window before executing anything else
[here (quite a messy commit)](https://github.com/end-4/dots-hyprland/commit/18f82b7a1f7bb98ea4d67d80a8447e44761568f9)

**Comment by D3vil0p3r** [comment]
> * I fixed the themer freezing issue by making those buttons close that window before executing anything else
> * Also created `_material_default.scss`
>   [here (quite a messy commit)](https://github.com/end-4/dots-hyprland/commit/18f82b7a1f7bb98ea4d67d80a8447e44761568f9)

The freezing issues about those two buttons seem to be fixed.

What I noted that sometimes, maybe due to a race condition, the same could occur also for "osettings" button (the one on the left of the top bar). Is it needed to use the same tactics?

I'm also noting that on the latest commit you wrote in your previous comment, it reverted the `allappnames.cpp` and `allapps.cpp` to the previous not working state, because I see that the check on "existing directory" has been removed (I'm not sure if also other files have been reverted to an old state). Was it intended or not?

**Comment by end-4** [comment]
### Buttons on sidebars
- Wifi settings: The config currently tries to open `gnome-control-center`, which doesn't seem to launch in Hyprland anymore... I'm not sure why. I'll fix that by replacing it with `foot` running `nmtui`, what do you think?
- Hypr Dev. Discord: I forgot `&&` in the command in the config. Oops.
- https://github.com/end-4/dots-hyprland/commit/d029bad933df33ba843c8a9ee76c63fad3b7b912 fixes the above two
- Lock: It does not need a specific dependency. I use a script for locking but forgot to include it.
- https://github.com/end-4/dots-hyprland/commit/8027ce6c7d360e7ff9048edf638615cd1ca50d4b fixes the above
### The app name/entry scripts
- It was my problem of copying-the-whole-folder-offline-and-upload. The effect was unintended.
- https://github.com/end-4/dots-hyprland/commit/9e965882c619ae1634c1d918d35935abe73985a7 fixes this.


I'm sorry for my irresponsibility, and thank you for your patience. I honestly admire you for your effort of digging into another person's configurations -- which has never been an easy job.

**Comment by D3vil0p3r** [comment]
**Buttons on sidebars**
* About `nmtui` I dont like because it does not fit well with the graphics of the theme. We should find something like gnome-control-center
* Lock works. Where can I specify the path where to set the lock background?
* Suspend button still does not work. Just flash the screen several times only

**The app name/entry scripts**
* Now I note that the icons of the applications on the app search don't appear anymore:
   ![image](https://github.com/end-4/dots-hyprland/assets/83867734/8da8abd5-10ad-4529-bb9b-8ffcf3ab25d7)

**Icon theme for Summer Gruv**
* About the icon theme you used in https://streamable.com/4oogot , which icon theme did you use for folders and files?

**Comment by end-4** [comment]
### Menu buttons
- About `nmtui`: I totally see your point. Which alternative do you recommend? I don't really mind as long as it works - it's not like you have to change wifi settings everyday...
- Lock: `gtklock` is used. Config is in `.config/gtklock/`, which includes the config, style css, and wallpaper.
- Suspend: The button just executes `systemctl suspend`. I don't think it's a problem related to my configs.
- Notifications: I see. Will add state change for that.
### Search icons
- You should not run `appicons`. An old script, I gotta clean that up. https://github.com/end-4/dots-hyprland/commit/4beaf28e65fdc78b5f5fbdff8f9e89df426a3a9b
- Delete files in `.config/eww/scripts/cache/` folder, open the overview, wait 5-10secs, close that and re-open it. Should work.
### Icon theme
- Still the win11 one.

**Comment by D3vil0p3r** [comment]
Sorry for being late. I tested the configuration now but:
* when I run this Hyprland config the first time, I get the following notification on top-right:
   ![image](https://github.com/end-4/dots-hyprland/assets/83867734/df7e47d7-eab3-4af3-8b9b-cee12b992289)
* Look the image above and the workspace numbers. I don't know why but now I have those two colors: that light red on 1 and yellow on 2. And now, if I try to switch to other workspaces by SUPER + 1 or 2 or 3 and so on, sometimes does not switch and I need to press that keybind several times.
* I tested that "pause" notification button, but sometimes it toggles to the play icon or toggles back to the pause icon only after several clicks and not by just one click.
* For app icons in the search bar, despite I have the correct icons in `.config/eww/scripts/cache/`, for example `kali-mimikatz-128x128.png`, that is associated to `.local/share/applications/mimikatz.desktop`, in the search bar it does not apply that icon to Mimikatz.
